File: WallClock/WallClock/ClockSliceView.m
//
//  ClockSliceView.m
//  WallClock
//
//  Created by Oliver Michalak on 16.07.17.
//  Copyright © 2017 Oliver Michalak. All rights reserved.
//

#import "ClockSliceView.h"
#import "ClockKit.h"

@implementation ClockSliceView

- (void)setDate:(NSDate *)date {
	_date = date;
	[self setNeedsDisplay];
}

- (void)setStartDate:(NSDate *)date {
	_startDate = date;
	[self setNeedsDisplay];
}

- (void)setEndDate:(NSDate *)date {
	_endDate = date;
	[self setNeedsDisplay];
}

- (BOOL)isPlaying {
	if (self.date && self.startDate && self.endDate) {
		// expensive but exiting out when minutes are same already
		NSDateComponents *dateComp = [[NSCalendar currentCalendar] components:NSCalendarUnitMinute fromDate:self.date];
		NSDateComponents *endComp = [[NSCalendar currentCalendar] components:NSCalendarUnitMinute fromDate:self.endDate];
		return (dateComp.minute != endComp.minute && [self.date compare:self.endDate] == NSOrderedAscending);
	}
	return NO;
}

// move to VC if LED clock should get timer too, (together with idleTimerDisabled)
- (void)togglePlayPause:(UITapGestureRecognizer *)gesture {
	if (gesture.state == UIGestureRecognizerStateEnded) {
		if (self.isPlaying) {
			self.startDate = self.endDate = nil;
		}
		else {
			self.startDate = [NSDate date];
			self.endDate = [self.startDate dateByAddingTimeInterval:60*15];	// 15mins slices
		}
	}
}

- (void)drawRect:(CGRect)rect {
	if (self.isPlaying) {
		self.hidden = NO;
		[UIApplication sharedApplication].idleTimerDisabled = YES;
		NSDateComponents *dateComp = [[NSCalendar currentCalendar] components:NSCalendarUnitMinute fromDate:self.date];
		NSDateComponents *endComp = [[NSCalendar currentCalendar] components:NSCalendarUnitMinute fromDate:self.endDate];

		// seconds precision preferred for coloring
		float diff = MAX(0, MIN(1, ([self.date timeIntervalSinceReferenceDate] - [self.startDate timeIntervalSinceReferenceDate]) / ([self.endDate timeIntervalSinceReferenceDate] - [self.startDate timeIntervalSinceReferenceDate])));
		// in RGB direct green->red creates ugly brown in the middle, just add blue in between
		UIColor *sliceColor = [UIColor colorWithRed:diff green:1-diff blue:(2*diff >= 1 ? 2 - 2*diff : 2*diff) alpha:1];

		[ClockKit drawClockSliceWithFrame:self.frame resizing:ClockKitResizingBehaviorAspectFit sliceColor:sliceColor sliceStartValue:dateComp.minute sliceEndValue:endComp.minute];
	}
	else {
		self.hidden = YES;
		[UIApplication sharedApplication].idleTimerDisabled = NO;
		self.startDate = self.endDate = nil;
	}
}

@end

File: WallClockSwift/WallClockSwift/ClockSliceView.swift
//
//  ClockSliceView.swift
//  WallClockSwift
//
//  Created by Oliver Michalak on 17.07.17.
//  Copyright © 2017 Oliver Michalak. All rights reserved.
//

import UIKit

class ClockSliceView: UIView {

	var date : Date? = nil {
		didSet {
			setNeedsDisplay()
		}
	}

	var startDate : Date? = nil {
		didSet {
			setNeedsDisplay()
		}
	}

	var endDate : Date? = nil {
		didSet {
			setNeedsDisplay()
		}
	}

	var isPlaying : Bool {
		guard let date = date,
					let endDate = endDate else { return false }
		let dateComp = Calendar.current.dateComponents([Calendar.Component.minute], from: date)
		let endComp = Calendar.current.dateComponents([Calendar.Component.minute], from: endDate)
		return dateComp.minute != endComp.minute && date.compare(endDate) == .orderedAscending
	}

	// move to VC if LED clock should get timer too, (together with idleTimerDisabled)
	public func togglePlayPause (gesture: UITapGestureRecognizer) {
		if (gesture.state == .ended) {
			if (isPlaying) {
				startDate = nil
				endDate = nil
			}
			else {
				startDate = Date()
				endDate = startDate?.addingTimeInterval(60*15)	// 15mins slices
			}
		}
	}

	override func draw(_ rect: CGRect) {
		if isPlaying {
			isHidden = false
			UIApplication.shared.isIdleTimerDisabled = true
			guard let date = date,
						let endDate = endDate,
						let startDate = startDate else { return }
			let dateComp = Calendar.current.dateComponents([Calendar.Component.minute], from: date)
			let endComp = Calendar.current.dateComponents([Calendar.Component.minute], from: endDate)

			// seconds precision preferred for coloring
			let diff = Float(fmax(0, fmin(1, (date.timeIntervalSinceReferenceDate - startDate.timeIntervalSinceReferenceDate) / (endDate.timeIntervalSinceReferenceDate - startDate.timeIntervalSinceReferenceDate))))
			// in RGB direct green->red creates ugly brown in the middle, just add blue in between
			let sliceColor = UIColor.init(colorLiteralRed: diff, green: 1-diff, blue: (2*diff >= 1 ? 2 - 2*diff : 2*diff), alpha: 1)
			ClockKit.drawClockSlice(frame: frame, resizing: .aspectFit, sliceColor: sliceColor, sliceStartValue: CGFloat(dateComp.minute!), sliceEndValue: CGFloat(endComp.minute!))
		}
		else {
			isHidden = true
			UIApplication.shared.isIdleTimerDisabled = false
			startDate = nil
			endDate = nil
		}
	}
}
